### ISSUE TITLE
Fan Speed Toggle

### ISSUE THREAD

**Issue by RobertN82** [kiwi-cam/homebridge-broadlink-rm#618]
**Describe the solution you'd like**
My fan has 3 speeds, and my remote only has one button that toggles to the next speed. Once it reaches max speed (3), it switches back to the slowest speed (1).

Some kind of speed toggle setting would be great

**Describe alternatives you've considered**
I tried setting it up with intervals so you can jump from speed 1 to 3, but this won't work as it needs relative intervals instead of absolutes. Not sure If it's possible for this reason?

**Comment by ggroel** [comment]
I have this configuration, in the home app, every time I decide at what speed, I turn off the fan and press the “speed up" button as many times as I chose at home. I hope it helps you,

                {
                    "name": "Ventilador Mir y Gastón",
                    "type": "fan",
                    "host": "10.0.0.22",
                    "hideRotationDirection": true,
                    "hideSwingMode": true,
                    "alwaysResetToDefaults": true,
                    "defaultFanSpeed": 33,
                    "stepSize": 33,
                    "data": {
                        "on": "2600500000012996121313131213131312131313121313131238133713371337133713371338123812381312133812131313121313371313121313371313133713371337131313371300052800012a4b13000d05000000000000",
                        "off": "2600500000012898101512141014121412141115121312131238123812381337123813381238123812381238133712141213131312381213131312141213123813371337121412381200052800012a4c12000d05000000000000",
                        "fanSpeed33": [
                            {
                                "data": "2600500000012898101512141014121412141115121312131238123812381337123813381238123812381238133712141213131312381213131312141213123813371337121412381200052800012a4c12000d05000000000000",
                                "pause": 1
                            },
                            {
                                "data": "2600500000012996121313131213131312131313121313131238133713371337133713371338123812381312133812131313121313371313121313371313133713371337131313371300052800012a4b13000d05000000000000"
                            }
                        ],
                        "fanSpeed66": [
                            {
                                "data": "2600500000012898101512141014121412141115121312131238123812381337123813381238123812381238133712141213131312381213131312141213123813371337121412381200052800012a4c12000d05000000000000",
                                "pause": 1
                            },
                            {
                                "data": "2600500000012996121313131213131312131313121313131238133713371337133713371338123812381312133812131313121313371313121313371313133713371337131313371300052800012a4b13000d05000000000000",
                                "sendCount": 2,
                                "interval": 1
                            }
                        ],
                        "fanSpeed99": [
                            {
                                "data": "2600500000012898101512141014121412141115121312131238123812381337123813381238123812381238133712141213131312381213131312141213123813371337121412381200052800012a4c12000d05000000000000",
                                "pause": 1
                            },
                            {
                                "data": "2600500000012996121313131213131312131313121313131238133713371337133713371338123812381312133812131313121313371313121313371313133713371337131313371300052800012a4b13000d05000000000000",
                                "sendCount": 3,
                                "interval": 1
                            }
                        ]
                    }
                }

**Comment by RobertN82** [comment]
Great idea but I think that only works if the fan resets after switching off?
Mine remembers the speed setting so I don't think this will work?

**Comment by ggroel** [comment]
I think that would have to set "alwaysResettodefaults": to false.

**Comment by RobertN82** [comment]
Let's try it and find out. Hope I'm wrong :-)

**Comment by RobertN82** [comment]
Ah no looking at your config it won't work for me. By the looks of it your power button has a different hex for and and off.
Mine toggles on/off with the same hex, so sending the on code multiple times won't work.

**Comment by ggroel** [comment]
No, in reality my On HEX does not correspond to my ON button, but to the speed button, since when pressing that, if the fan is off turns on it.

**Comment by RobertN82** [comment]
Ah I get it. I tried it with my remote, but sadly mine fan does not turn on with the speed button. It needs to be turned on first, only then does it accept speed commands.

**Comment by ggroel** [comment]
Then your HEX for ON will be different from the HEX to speed up

**Comment by RobertN82** [comment]
Correct, they are different.
I only have a hex for power on/off (same) and a hex for speed up.

Without it resetting to 33% each time on startup or storing it's last known state, I'm not really sure if it can be done?

**Comment by RobertN82** [comment]
I think I've got it going now with setting it 2 speed steps (speed 1 and 3).

```
{
            "name": "Fan",
            "type": "fan",
            "host": "...",
            "hideRotationDirection": true,
            "alwaysResetToDefaults": false,
            "defaultFanSpeed": 100,
            "stepSize": 50,
            "data": {
                "on": "...", //power
                "off": "...", //power
                "swingToggle": "...",
                "fanSpeed50": [
                    {
                        "data": "..."  //speed
                    }
                ],
                "fanSpeed100": [
                    {
                        "data": "...", //speed
                                "sendCount": 2,
                                "interval": 1
                    }
                ]
            }
        }
```

**Comment by ggroel** [comment]
Try this, I think it will work ... I changed the parameter "AlwaysSetTodefaults" and deleted the “DefaultFanSpeed" parameter so that I never sent a default value. On the other hand, with each speed change you have to turn off the fan, turn it on and set the speed...

{
                    "name": "Ventilador Mir y Gastón",
                    "type": "fan",
                    "host": "10.0.0.22",
                    "hideRotationDirection": true,
                    "hideSwingMode": true,
                    "alwaysResetToDefaults": false,
                    "stepSize": 33,
                    "data": {
                        "on": "ON/OFF HEX",
                        "off": "ON/OFF HEX",
                        "fanSpeed33": [
                            {
                                "data": "ON/OFF HEX”,
                                "sendCount": 2,
                                "interval": 1
                            },
                            {
                                "data": "SPEED UP HEX"
                            }
                        ],
                        "fanSpeed66": [
                            {
                                "data": "ON/OFF HEX”,
                                "sendCount": 2,
                                "interval": 1
                            },
                            {
                                "data": "SPEED UP HEX",
                                "sendCount": 2,
                                "interval": 1
                            }
                        ],
                        "fanSpeed99": [
                            {
                                "data": "ON/OFF HEX”,
                                "sendCount": 2,
                                "interval": 1
                            },
                            {
                                "data": "SPEED UP HEX",
                                "sendCount": 2,
                                "interval": 1
                            }
                        ]
                    }
                }

**Comment by RobertN82** [comment]
Haven't had a chance to try your latest config, but I think it won't work going from speed 2 back to speed 1? 
This is why I just set it up as a 2 speed (1 and 3) fan which seems to work quite well now.
